pallets/sponsorship/src/migration.rs

```diff
@@ -0,0 +1,342 @@
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+use crate::{
+	BalanceOf, Config, Pallet, Pot, PotDetailsOf, PotMigrationCursor, User, UserDetailsOf, UserMigrationCursor,
+	UserRegistrationCount,
+};
+use codec::{Decode, Encode};
+use frame_support::{
+	pallet_prelude::*,
+	storage::generator::{StorageDoubleMap, StorageMap},
+	traits::{Get, StorageVersion},
+	weights::Weight,
+};
+use support::LimitedBalance;
+pub use v0::migrate_partially;
+
+/// The current storage version.
+pub const STORAGE_VERSION: StorageVersion = StorageVersion::new(1);
+
+mod v0 {
+	use super::{Pot as V1Pot, PotDetailsOf as V1PotDetailsOf, User as V1User, UserDetailsOf as V1UserDetailsOf, *};
+	use frame_support::storage_alias;
+	use sp_runtime::traits::Saturating;
+
+	#[derive(Encode, Decode, Debug)]
+	pub struct PotDetails<AccountId, SponsorshipType, Balance: frame_support::traits::tokens::Balance> {
+		pub sponsor: AccountId,
+		pub sponsorship_type: SponsorshipType,
+		pub fee_quota: LimitedBalance<Balance>,
+		pub reserve_quota: LimitedBalance<Balance>,
+	}
+
+	#[derive(Encode, Decode, Debug)]
+	pub struct UserDetails<AccountId, Balance: frame_support::traits::tokens::Balance> {
+		pub proxy: AccountId,
+		pub fee_quota: LimitedBalance<Balance>,
+		pub reserve_quota: LimitedBalance<Balance>,
+	}
+
+	pub type PotDetailsOf<T> =
+		PotDetails<<T as frame_system::Config>::AccountId, <T as Config>::SponsorshipType, BalanceOf<T>>;
+	pub type UserDetailsOf<T> = UserDetails<<T as frame_system::Config>::AccountId, BalanceOf<T>>;
+
+	#[storage_alias]
+	/// Details of a pot.
+	type Pot<T: Config> = StorageMap<Pallet<T>, Blake2_128Concat, <T as Config>::PotId, PotDetailsOf<T>, OptionQuery>;
+
+	#[storage_alias]
+	/// User details of a pot.
+	type User<T: Config> = StorageDoubleMap<
+		Pallet<T>,
+		Blake2_128Concat,
+		<T as Config>::PotId,
+		Blake2_128Concat,
+		<T as frame_system::Config>::AccountId,
+		UserDetailsOf<T>,
+		OptionQuery,
+	>;
+
+	fn migrate_pot_partially<T: Config>(n: usize, starting_key: Vec<u8>) -> (Option<Vec<u8>>, Weight) {
+		let mut iter = Pot::<T>::iter_from(starting_key);
+
+		let pots = iter
+			.by_ref()
+			.take(n)
+			.map(|(pot, details)| {
+				(
+					pot,
+					V1PotDetailsOf::<T> {
+						sponsor: details.sponsor,
+						sponsorship_type: details.sponsorship_type,
+						fee_quota: details.fee_quota,
+						reserve_quota: details.reserve_quota,
+						deposit: Default::default(),
+					},
+				)
+			})
+			.collect::<Vec<_>>();
+
+		let num_of_pots = pots.len();
+
+		pots.into_iter()
+			.for_each(|(pot, details)| V1Pot::<T>::insert(pot, details));
```

Comment:
Can't this be done with only one iterator?

Reply:
The thing that might not be obvious here is that v0::Pot and V1Pot are both referencing the same storage. We shouldn't make any changes to one of them while iterating over another one. So I need to read from Pot, do my operation and make it sure it's applied (by collecting it) so I can then safely start to iterate and change the other one. Any suggestion is welcomed.